ToSchema class with Generic deriving and other helpers

fizruk opened the pull request:
What's in:
- [x] `ToSchema` class with default `Generic`-based implementation;
- [x] `genericToSchema` working for some base types and record data types;
- [x] `toSchemaBoundedIntegral` and `toSchemaBoundedEnum` helpers;
- [x] introduce `Options` similar to `Data.Aeson.Options` (I have added `fieldLabelModifier`, what else do we need?);
- [x] more tests/examples;
- [x] more `ToSchema` instances for common types;
- [x] add `ToSchema` instances for `time` types;
- [ ] `ToParam`/`ToHeader` classes;
- [x] documentation.

@dmjio would this be useful for `servant-swagger`?

dmjio commented:
@fizruk, fantastic work, this is definitely useful for `servant-swagger`, in fact this should make `servant-swagger` trivial to port now. 

I think putting the generics based derivations in `swagger2` makes more sense. I'm also a fan of using `GHC.Generics` as opposed to another generics library, since we incur no additional dependency cost. (And I think all other generics libraries rely on GHC.Generics anyways, since it's caked into `GHC` IIRC) 

 Still wondering if `ToParam` and `ToHeader` would be better suited for `servant-swagger` or `swagger2`. I think a case could be made for both. Either way I think it would be relatively simple to add. 

Overall, great work :+1:

fizruk commented:
Just to keep you in loop, in recent commits I have
- changed `toSchema` to `toNamedSchema` to enable `$ref`s;
- added `Generic` deriving for sum types (following [`ObjectWithSingleField` sum encoding](https://hackage.haskell.org/package/aeson-0.10.0.0/docs/Data-Aeson-Types.html#t:SumEncoding));
- added `Generic` deriving for "all nullary" data types and removed "bounded enum" helpers;
- added support for unit types (e.g. `()`) following `aeson` which encodes unit type as `[]`;
- added a few more options to `SchemaOptions`:
    - `constructorTagModifier`;
    - `datatypeNameModifier`;
    - `allNullaryToStringTag`;
- added more example tests;
- fixed a few things;

fizruk commented:
This PR has grown big and before I start to change things even further I think it is time to merge and release.

I still am a bit uncomfortable with `NamedSchema` and `useReferences` though.
The major problem I think is that when building the entire `Swagger` spec we can't know for sure all needed schema definitions for a spec to be complete. Here's an example:

```haskell
data Color = White | Black deriving (Generic, ToSchema)
data Dog = Dog { nick :: String, color :: Color } deriving (Generic, ToSchema)
data Person = Person { name :: String, dog :: Dog } deriving (Generic, ToSchema)

type PersonAPI
  = "person" :> ReqBody Person :> Post ()
```

How would you know to add `Dog` and `Color` to auto-generated swagger spec?

The solution I think should be for `ToSchema` to generate also a list of used definitions.
But as I have already mentioned, I would like to merge and release current changes before moving forward.

@dmjio what do you think?

dmjio commented:
:+1:
